tools/codegen/src/main.rs

```diff
@@ -287,19 +285,40 @@ fn main() -> Result<()> {
                 }
             };
 
-            eprint!("downloading {url} for checksum ... ");
+            eprint!("downloading {url} for etag and checksum ... ");
             let download_cache = &download_cache_dir.join(format!(
                 "{version}-{platform:?}-{}",
                 Path::new(&url).file_name().unwrap().to_str().unwrap()
             ));
+            let response = download(&url)?;
+            let etag = response
+                .header("etag")
+                .expect("binary should have an etag")
+                .to_string()
+                .replace('\"', "");
+
+            if let Some(ManifestRef::Real(ref manifest)) = existing_manifest {
+                if let Some(entry) = manifest.download_info.get(&platform) {
+                    if entry.etag == etag {
```

Comment:
According to [mdn doc for etag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#caching_of_unchanged_resources), we could include `None-Match: "${etag}"` in the request header, and the server will return [304](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304) Not Modified status, without a body.

Reply:
Aha sorry I didn't read the description

> Using If-None-Match header would also work, but that would add complexity to download(..)

Hmmm that's true... Though I think we can create a new `download_if_not_cached` function to reduce the code complexity?

But if you think it's too much work, feel free to close this one, reducing 4m to 2m is already a lot.

> I found using head doesnt work, as the Server responded to ureq::head with 401 if If-None-Match is used; however responds to wget head with 304.

Perhaps the web server reacts differently to different user agent.

> as the asset fetches because they are so numerous and voluminous to be likely to prevent features being developed by new contributors who will balk at downloading so much data just to check a new feature against the full set of manifests.

I think the refetching is done to verify the data, which makes sense, though does make CI longer.

One thing we could do with that, is to checked for yank crates and remove them from the action.

Reply:
`download_if_not_cached` would be a lot of duplication.  And because github assets dont allow HEAD and etags with the ureq default user agent, we would need to make it github specific.  Might be worth it if we needed that, but we dont need it for assets.  This PR is effectively doing a HEAD and disconnecting if the etag in the header matches, before streaming the asset body.  We are doing the `If-None-Match` on the client side.

Removing yanked crates from the manifest is a great idea - but it is orthogonal to this feature, so should be a new issue.

CI only refreshes the latest version, to validate them.